### PR TITLE
Delegate parameter parsing to specialised classes

### PR DESCRIPTION
**Pull request by LorenzoCelli**
Issue #38

### CODE REVIEW

File: tool2schema/parameter_schema.py [hunk 1]
```diff
@@ -0,0 +1,186 @@
...
+    """
+    Base class for typing module parameter schemas (e.g., typing.List).
+    """
+
+    def get_sub_type(self):
+        if "__args__" in dir(self.parameter.annotation):
+            annotation_name = self.parameter.annotation.__args__[0].__name__
+            return TYPE_MAP.get(annotation_name, "object")
+
+        return None
```
Comment (siliconlad) [inline tool2schema/parameter_schema.py:141]: Can this not be replaced with `typing.get_args`?

File: tool2schema/parameter_schema.py [hunk 2]
```diff
@@ -0,0 +1,186 @@
...
+class TypingOptionalParameterSchema(TypingParameterSchema):
+    """
+    Parameter schema for typing.Optional types.
+    """
+
+    @staticmethod
+    def matches(parameter: Parameter) -> bool:
+        return (
+            parameter.annotation != parameter.empty
+            and re.match(r"typing\.Optional.*", str(parameter.annotation)) is not None
```
Comment (siliconlad) [inline tool2schema/parameter_schema.py:173]: Could we potentially replace this with `typing.get_origin`?

**Review by siliconlad — CHANGES_REQUESTED** [review]
I'm also a fan of prepending all the internal methods with `_`.

For example `_add_type` and `_add_description`.

So it's clear that `to_json` and `matches` is the public facing API.

**Comment by siliconlad** [comment]
Thanks to you I've discovered all the methods that exists within `typing` like `get_args` and `get_origin` which I think should simplify a lot of the code I've written with regex 👍

**Comment by LorenzoCelli** [comment]
Should be done. Using `typing.get_origin` and `typing.get_args` I was able to remove `TypingListParameterSchema` and now `ListParameterSchema` handles both `list` and `typing.list`. I renamed `TypingOptionalParameterSchema` to `OptionalParameterSchema` for uniformity.

File: tool2schema/parameter_schema.py [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+class ListParameterSchema(GenericParameterSchema):
+    """
+    Parameter schema for list (array) types.
+    """
+
+    @staticmethod
+    def matches(parameter: Parameter) -> bool:
+        return parameter.annotation != Parameter.empty and (
+            parameter.annotation is list
+            or typing.get_origin(parameter.annotation) is list
```
Comment (siliconlad) [inline tool2schema/parameter_schema.py:132]: I believe `typing.get_origin` works for both `list` and `typing.List`.

File: tool2schema/parameter_schema.py [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+    """
+    Parameter schema for typing.Optional types.
+    """
+
+    @staticmethod
+    def matches(parameter: Parameter) -> bool:
+        return (
+            parameter.annotation != parameter.empty
+            and typing.get_origin(parameter.annotation) is typing.Union
+            and type(None) in typing.get_args(parameter.annotation)
```
Comment (siliconlad) [inline tool2schema/parameter_schema.py:153]: Do you think checking length of args here is important, given `typing.Optional` accepts only one type.

**Review by siliconlad — CHANGES_REQUESTED** [review]
Liking the look of this!

File: tool2schema/parameter_schema.py (hunk 3 above)
Comment (LorenzoCelli) [reply tool2schema/parameter_schema.py:132]: Unfortunately it returns `None` when the type hint is `list` without subscription :(

File: tool2schema/parameter_schema.py (hunk 4 above)
Comment (LorenzoCelli) [reply tool2schema/parameter_schema.py:153]: Not sure I understand the question. In this case typing.get_args returns a tuple with two elements, because it sees the optional as a union `(None, Y)` where Y is some other type

Comment (LorenzoCelli) [reply tool2schema/parameter_schema.py:153]: Or are you referring to the case when the optional is not subscripted?

Comment (LorenzoCelli) [reply tool2schema/parameter_schema.py:153]: Do we want to support it? It would be like a parameter without a type hint
```python
# a is whatever
def f(a: typing.Optional):
    pass
```

File: tool2schema/parameter_schema.py (hunk 3 above)
Comment (siliconlad) [reply tool2schema/parameter_schema.py:132]: Sad :(

File: tool2schema/parameter_schema.py (hunk 4 above)
Comment (siliconlad) [reply tool2schema/parameter_schema.py:153]: I don't we need to support that.

Sorry, let me rephrase my original question (which wasn't very clear 😓)

Should be check if the number of args is 2? Because a user could conceivably do

```python
def f(a: typing.Union[str, list, None])
```

We then would assume that it was an optional type of `str`.

Comment (LorenzoCelli) [reply tool2schema/parameter_schema.py:153]: Got it, will do later

Comment (LorenzoCelli) [reply tool2schema/parameter_schema.py]: Done

File: tool2schema/parameter_schema.py [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+    def _add_type(self, schema: dict):
+        if sub_type := super()._get_sub_type():
+            schema["type"] = sub_type
+
+
+PARAMETER_SCHEMAS = [
+    OptionalParameterSchema,
+    ListParameterSchema,
+    ValueTypeSchema,
+]
```
Comment (siliconlad) [inline tool2schema/parameter_schema.py:167]: Maybe just add a comment here saying the order is important? 

Because things break if you defined `PARAMETER_SCHEMAS` like

```python
PARAMETER_SCHEMAS = [
    ValueTypeSchema,
    OptionalParameterSchema,
    ListParameterSchema,
]
```

Because every `list` would match with `ValueTypeSchema` first.

**Review by siliconlad — CHANGES_REQUESTED** [review]
Last thing!

File: tool2schema/parameter_schema.py (hunk 5 above)
Comment (LorenzoCelli) [reply tool2schema/parameter_schema.py:170]: Added! 😄

**Review by siliconlad — APPROVED** [review]
LGTM!
